sdk/Lusid.Sdk/Utilities/ClientCredentialsFlowTokenProvider.cs

```diff
@@ -63,26 +67,47 @@ public ClientCredentialsFlowTokenProvider(ApiConfiguration configuration)
         /// <inheritdoc />
         public async Task<string> GetAuthenticationTokenAsync()
         {
-            await _semaphore.WaitAsync();
-            try
+            AsyncRetryPolicy _policy;
+            _policy = 
+                Policy
+                    .Handle<HttpRequestException>()
+                    .WaitAndRetryAsync(5, retryAttempt =>
+                        TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)), OnRetry);
+            
+            var r = await _policy.ExecuteAsync(context => GetAuthenticationTokenAsyncInternal(), CancellationToken.None);
+            return r;
+
+            async Task<string> GetAuthenticationTokenAsyncInternal()
             {
-                if (_lastIssuedToken == null || _lastIssuedToken.ExpiresOn < DateTimeOffset.UtcNow)
+                await _semaphore.WaitAsync();
+                try
                 {
-                    if (_lastIssuedToken?.RefreshToken != null)
+                    if (_lastIssuedToken == null || _lastIssuedToken.ExpiresOn < DateTimeOffset.UtcNow)
                     {
-                        _lastIssuedToken = await RefreshToken(_apiConfig, _lastIssuedToken.RefreshToken);
-                    }
-                    else
-                    {
-                        _lastIssuedToken = await GetNewToken(_apiConfig);
+                        if (_lastIssuedToken?.RefreshToken != null && _lastIssuedToken?.RefreshExpiresOn > DateTimeOffset.UtcNow)
+                        {
+                            _lastIssuedToken = await RefreshToken(_apiConfig, _lastIssuedToken?.RefreshToken);
+                        }
+                        else
+                        {
+                            _lastIssuedToken = await GetNewToken(_apiConfig);
+                        }
                     }
                 }
+                finally
+                {
+                    _semaphore.Release();
+                }
+                return _lastIssuedToken.Token;
             }
-            finally
+        }
+
+        private void OnRetry(Exception arg1, TimeSpan arg2)
+        {
+            if (arg1.Message.Contains("refresh token is invalid or expired"))
```

Comment:
it would be more robust if this was based on the http status code, presumably anything other that a 200 range response should be considered a failure

Reply:
if we did this for every error then it would have to get a whole new token every time instead of using the refresh token. I don't think this behavior would be wanted for something like a 429?

Reply:
if we are doing string comparison, I would also make it lowercase to account for any rogue capitalisation